proposals/roles.md

```diff
@@ -0,0 +1,704 @@
+# Extension types
+
+## Summary
+[summary]: #summary
+
+The purpose of "extensions" is to augment or adapt existing types to new scenarios, when those types are not under your control, or where changing them would negatively impact other uses of them. The adaptation can be in the form of adding new function members as well as implementing additional interfaces.
+
+## Motivation
+[motivation]: #motivation
+
+Explicit extensions address two main classes of scenarios: "augmentation" scenarios (fit an existing value with a *new* member) and "adaptation" scenarios (fit an existing value to an *existing* interface). 
+
+In addition, implicit extensions would provide for additional kinds of extension members beyond today's extension methods, and for "extension interfaces".
+
+### Augmentation scenarios
+
+Augmentation scenarios allow existing values to be seen through the lens of a "stronger" type - an extension that provides additional function members on the value.
+
+### Adaptation scenarios
+
+Adaptation scenarios allow existing values to be adapted to existing interfaces, where an extension provides details on how the interface members are implemented.
+
+## Design
+
+This proposal is divided into three parts, all relating to extending existing types:  
+A. static implicit extensions  
+B. implicit and explicit extensions with members  
+C. implicit and explixit extensions that implement interfaces  
+
+The syntax for extensions is as follows:
+
+```antlr
+type_declaration
+    | extension_declaration // add
+    | ...
+    ;
+
+extension_declaration
+    : extension_modifier* 'extension' identifier type_parameter_list? 'for' type extension_base_type_list? type_parameter_constraints_clause* extension_body
+    ;
+
+extension_base_type_list
+    : ':' extension_or_interface_type_list
+    ;
+
+extension_or_interface_type_list
+    : interface_type_list
+    : extension_type (',' extension_or_interface_type_list)
+    ;
+
+extension_body
+    : '{' extension_member_declaration* '}'
+    ;
+
+extension_member_declaration
+    : constant_declaration
+    | field_declaration
+    | method_declaration
+    | property_declaration
+    | event_declaration
```

Comment:
events also seem problematic.  or is this just add/remove events, and not field-style events?